src/libraries/System.IO.FileSystem.AccessControl/src/System/IO/MemoryMappedFiles/MemoryMappedFileAcl.cs

```diff
@@ -0,0 +1,205 @@
+        /// The default is <see cref="HandleInheritability.None" />.</param>
+        /// <param name="leaveOpen">This parameter is unused: the <paramref name="fileStream" /> will be left open after the current object is disposed.</param>
+        /// <returns>A memory-mapped file that has the specified characteristics.</returns>
+        /// <exception cref="ArgumentNullException"><paramref name="fileStream" /> or <paramref name="memoryMappedFileSecurity" /> is <see langword="null" />.</exception>
+        /// <exception cref="ArgumentException"><paramref name="mapName" /> is <see langword="null" /> or empty.
+        /// -or-
+        /// The file is empty: <paramref name="capacity" /> is 0 and the length of <paramref name="fileStream" /> is 0.
+        /// -or-
+        /// <paramref name="access"/> is <see cref="MemoryMappedFileAccess.Write" />.
+        /// -or-
+        /// <paramref name="access"/> is <see cref="MemoryMappedFileAccess.Read" /> and the <paramref name="capacity" /> is larger than the length of <paramref name="fileStream" />.</exception>
+        /// <exception cref="ArgumentOutOfRangeException"><paramref name="capacity" /> is negative.
+        /// -or-
+        /// <paramref name="access" /> or <paramref name="inheritability" /> is not a valid enumeration value.
+        /// -or-
+        /// The capacity may not be smaller than the file size.</exception>
+        /// <remarks>
+        /// If `capacity` is larger than the size of the file on disk, the file on disk is increased to match the specified capacity even if no data is written to the memory-mapped file. To prevent this from occurring, specify 0 (zero) for the default capacity, which will internally set `capacity` to the size of the file on disk.
+        /// </remarks>
+        public static unsafe MemoryMappedFile CreateFromFile(
+            FileStream fileStream,
+            string? mapName,
+            long capacity,
+            MemoryMappedFileAccess access,
+            MemoryMappedFileSecurity memoryMappedFileSecurity,
+            HandleInheritability inheritability,
+            bool leaveOpen)
+        {
+            long updatedCapacity = MemoryMappedFileInternal.VerifyParametersCreateFromFile(fileStream, mapName, capacity, access, inheritability);
+
+            if (memoryMappedFileSecurity == null)
+            {
+                throw new ArgumentNullException(nameof(memoryMappedFileSecurity));
+            }
+
+            fixed (byte* pSecurityDescriptor = memoryMappedFileSecurity.GetSecurityDescriptorBinaryForm())
+            {
+                SafeMemoryMappedFileHandle handle = MemoryMappedFileInternal.CreateCore(fileStream, mapName, access, MemoryMappedFileOptions.None, updatedCapacity, MemoryMappedFileInternal.GetSecAttrs(pSecurityDescriptor, inheritability));
+                return new MemoryMappedFile(handle); // This constructor leaves the handle open
```

Comment:
The user will have to take care of manually disposing the passed `FileStream` after this `MemoryMappedFile` object is disposed.

Reply:
Note: The `MemoryMappedFile(SafeMemoryMappedFileHandle, FileStream, Boolean)` is not being used right now because it wasn't approved in the API Proposal meeting. It's what we currently [use in Framework](https://referencesource.microsoft.com/#System.Core/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,244), and what we use in the [existing .NET Core method](https://source.dot.net/#System.IO.MemoryMappedFiles/System/IO/MemoryMappedFiles/MemoryMappedFile.cs,254).

After discussing this with @JeremyKuhne and @ericstj, there seem to be enough reasons to go to the API Proposal meeting again to request making this constructor public - It may be an ugly pattern, but we have precedent of other classes using it, and we don't have a better choice 
 - we should not use reflection to call the private constructor, and we should not leave a fileStream open even if the user indicated they want it closed.

Other examples of public constructors that use a similar pattern:

- https://source.dot.net/#System.Private.CoreLib/BinaryReader.cs,49
- https://source.dot.net/#System.Private.CoreLib/BinaryWriter.cs,51

cc @bartonjs @stephentoub @GrabYourPitchforks